### PR TITLE
Tracked resource validation linter rules

### PR DESCRIPTION
**Pull request by sarangan12**
@dsgouda @vishrutshah @veronicagg @salameer Please review and approve

### CODE REVIEW

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 1]
```diff
@@ -253,6 +253,18 @@ Modelers:
...
+  <data name="TrackedResourceListByImmediateParentOperationMissing" xml:space="preserve">
+    <value>The child tracked resource, "{0}" with immediate parent "{1}", must have a list by immediate parent operation.</value>
+  </data>
+  <data name="TrackedResourceListByResourceGroupOperationMissing" xml:space="preserve">
+    <value>The tracked resource, "{0}", must have a list by resource group operation.</value>
+  </data>
+  <data name="TrackedResourceListBySubscriptionsOperationMissing" xml:space="preserve">
+    <value>The tracked resource, "{0}", must have a list by subscriptions operation.</value>
+  </data>
+  <data name="TrackedResourcePropertiesValidation" xml:space="preserve">
```
Comment (dsgouda) [inline src/core/AutoRest.Core/Properties/Resources.resx:265]: How is this different than https://github.com/Azure/autorest/issues/1967

File: src/modeler/AutoRest.Swagger/Model/ServiceDefinition.cs [hunk 2]
```diff
@@ -109,13 +109,17 @@ public ServiceDefinition()
...
         [Rule(typeof(ResourceModelValidation))]
-        [Rule(typeof(TrackedResourceValidation))]
-        [Rule(typeof(TrackedResourcePatchOperationValidation))]
-        [Rule(typeof(TrackedResourceGetOperationValidation))]
         [Rule(typeof(ResourceIsMsResourceValidation))]
         [Rule(typeof(GuidValidation))]
         [Rule(typeof(SkuModelValidation))]
         [Rule(typeof(DefinitionsPropertiesNamesCamelCase))]
+        [Rule(typeof(TrackedResourceGetOperationValidation))]
+        [Rule(typeof(TrackedResourceListByImmediateParent))]
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Model/ServiceDefinition.cs:117]: Question: Are we checking for "ANY" of (list by immediate parent, list by resource group and list by subscriptions) or "ALL" of them

File: src/modeler/AutoRest.Swagger/Model/ValidationUtilities.cs [hunk 3]
```diff
@@ -404,6 +404,24 @@ public static bool IsReferenceOfType(string reference, Dictionary<string, Schema
             return false;
         }
 
+        public static bool ListByXCheck(IEnumerable<Operation> getOperations, Regex regEx, string definitionKey, Dictionary<string, Schema> definitions)
+        {
+            return getOperations.Any(operation =>
+                       regEx.IsMatch(operation.OperationId) &&
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Model/ValidationUtilities.cs:410]: Simplify this a little
```getOperations.Where(op=>regEx.IsMatch(operation.OperationId) && ...).Any()```
Also, ValidationUtilities already has methods each to check for xmspageable and arraytype responses, please leverage those

File: src/modeler/AutoRest.Swagger/Validation/DescriptionMissing.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        /// Validates model for description property
+        /// </summary>
+        /// <param name="schema">Schema being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <returns><c>true</c> if model contains description, <c>false</c> otherwise</returns>
+        public override IEnumerable<ValidationMessage> GetValidationMessages(Dictionary<string, Schema> definitions, RuleContext context)
+        {
+            foreach (KeyValuePair<string, Schema> definition in definitions)
+            {
+                string key = definition.Key;
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/DescriptionMissing.cs:25]: nit: Or you could simply say
```
if (string.IsNullOrWhiteSpace(definition.Value.Description))
{
          yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, string.Format(ModelTypeFormatter, definition.Key));
}
```

File: src/modeler/AutoRest.Swagger/Validation/DescriptiveDescriptionRequired.cs [hunk 5]
```diff
@@ -27,7 +27,7 @@ internal static bool IsImpermissibleValue(this string description)
         }
     }
 
-    public class DescriptiveDescriptionRequired : TypedRule<string>
+    public class DescriptiveDescriptionRequired : DescriptionRequired<string>
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/DescriptiveDescriptionRequired.cs:30]: So.... no IsValid or GetValidationMessages implementations here?

File: src/modeler/AutoRest.Swagger/Validation/Extensions/LongRunningResponseValidationRule.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+        /// <summary>
+        /// The severity of this message (ie, debug/info/warning/error/fatal, etc)
+        /// </summary>
+        public override Category Severity => Category.Error;
+
+        /// <summary>
+        /// The template message for this Rule.
+        /// </summary>
+         public override string MessageTemplate => Resources.LongRunningResponseNotValid;
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/Extensions/LongRunningResponseValidationRule.cs:40]: nit: format

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs [hunk 7]
```diff
@@ -1,34 +1,24 @@
...
-        /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("Operation '{0}'", entity.OperationId) };
+            return !string.IsNullOrWhiteSpace(entity?.Description);
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs:21]: At this point, ```entity``` cannot be null, so remove the ```?```

File: src/modeler/AutoRest.Swagger/Validation/TrackedResourceGetOperationValidation.cs [hunk 8]
```diff
@@ -39,23 +37,27 @@ public class TrackedResourceGetOperationValidation : TypedRule<Dictionary<string
...
+            IEnumerable<string> trackedResources = context.TrackedResourceModels;
+
+            // Retrieve the list of getOperations
+            IEnumerable<Operation> getOperations = ValidationUtilities.GetOperationsByRequestMethod("get", context.Root);
+
+            foreach (string trackedResource in trackedResources)
             {
-                if (context.TrackedResourceModels.Contains(definition.Key.StripDefinitionPath()))
+                // check for 200 status response models since they correspond to a successful get operation
+                if (!getOperations.Any(op => op.Responses.ContainsKey("200") && (op.Responses["200"]?.Schema?.Reference?.StripDefinitionPath()).Equals(trackedResource)))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/TrackedResourceGetOperationValidation.cs:51]: What if this ```op.Responses["200"]?.Schema?.Reference?.StripDefinitionPath()``` evaluates to null

File: src/modeler/AutoRest.Swagger/Validation/TrackedResourceListByImmediateParent.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+        /// <summary>
+        /// Get the paths that fits the child resource criteria.
+        /// </summary>
+        /// <param name="childResource">name of the child resource</param>
+        /// <param name="serviceDefinition">service definition</param>
+        /// <returns>the paths that fit the child resource criteria</returns>
+        private IEnumerable<KeyValuePair<string, Dictionary<string, Operation>>> GetChildResourcesPaths(string childResource, ServiceDefinition serviceDefinition)
+        {
+            Regex pathRegEx = new Regex("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/.+/.+/{.+}.*/" + childResource + "s$", RegexOptions.IgnoreCase);
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/TrackedResourceListByImmediateParent.cs:46]: Use this regex instead ```Subscriptions\/\{.+\}\/ResourceGroups\/\{.+\}\/providers\/([^\/]+)(?<typenamesandvalues>[^?]+)([\?]?.+)?```

**Review by dsgouda — CHANGES_REQUESTED** [review]
More comments on the way

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs (hunk 7 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs:21]: Done

File: src/modeler/AutoRest.Swagger/Validation/Extensions/LongRunningResponseValidationRule.cs (hunk 6 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Validation/Extensions/LongRunningResponseValidationRule.cs:40]: Done

File: src/modeler/AutoRest.Swagger/Validation/DescriptiveDescriptionRequired.cs (hunk 5 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Validation/DescriptiveDescriptionRequired.cs:30]: Yes. It is in the base class

File: src/modeler/AutoRest.Swagger/Validation/DescriptionMissing.cs (hunk 4 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Validation/DescriptionMissing.cs:25]: Done

File: src/modeler/AutoRest.Swagger/Model/ValidationUtilities.cs (hunk 3 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Model/ValidationUtilities.cs:410]: Done

File: src/core/AutoRest.Core/Properties/Resources.resx (hunk 1 above)
Comment (sarangan12) [reply src/core/AutoRest.Core/Properties/Resources.resx:265]: Removed it

File: src/modeler/AutoRest.Swagger/Validation/TrackedResourceGetOperationValidation.cs (hunk 8 above)
Comment (sarangan12) [reply src/modeler/AutoRest.Swagger/Validation/TrackedResourceGetOperationValidation.cs:51]: redesigned it

**Pull request closed without merging**